R/fotmob_stats.R

```diff
@@ -0,0 +1,215 @@
+
+## TODO: Cache this
+.fotmob_get_season_ids <- function(...) {
+  .fotmob_get_league_or_season_ids(
+    url = "https://raw.githubusercontent.com/JaseZiv/worldfootballR_data/466d399903d2dfc2c0e37058dc2e3f33f1ac74d1/raw-data/fotmob-leagues/season_ids.csv",
+    ...
+  )
+}
+
+## Probably best to just have this always available, until caching is implemented.
+.fotmob_stat_types <- read.csv(
+  "https://raw.githubusercontent.com/JaseZiv/worldfootballR_data/bdc5a8f8d4b31b477621e309ddb06fae9d35e022/raw-data/fotmob-stats/stat_types.csv",
+  stringsAsFactors = F
+)
+
+#' @importFrom dplyr filter
+.fotmob_validate_stat_type <- function(team_or_player = c("player", "team"), stat_type = NULL) {
+
+  stopifnot(
+    "`stat_type` must be specified." = !is.null(stat_type),
+    "`stat_type` must have length 1" = length(stat_type) == 1
+  )
+
+  entity <- match.arg(team_or_player)
+  filt_stat_choices <- .fotmob_stat_types %>%
+    dplyr::filter(.data$entity == !!team_or_player)
+
+  n_stats <- length(stat_type)
+  res <- filt_stat_choices %>%
+    dplyr::filter(
+      .data$stat == !!stat_type
+    )
+  n_res <- nrow(res)
+
+  if(n_res == 0) {
+    stop(
+      "Could not find any matching `stat_type`."
+    )
+  }
+  res
+}
+
+#' @importFrom jsonlite fromJSON
+#' @importFrom tibble as_tibble
+#' @importFrom dplyr select
+#' @importFrom tidyr unnest
+#' @importFrom janitor clean_names
+#' @importFrom purrr safely
+.fotmob_get_single_season_stats <- function(
+  league_id,
+  season_id,
+  team_or_player = NULL,
+  stat_type = NULL
+) {
+
+  stat_type_df <- .fotmob_validate_stat_type(
+    team_or_player = team_or_player,
+    stat_type = stat_type
+  )
+
+  url <- sprintf(
+    "https://data.fotmob.com/stats/%s/season/%s/%s.json",
+    league_id,
+    season_id,
+    stat_type_df$full_stat
+  )
+
+  ## This still print out HTTP error 403 even with `quiet = TRUE`?!?
+  fs <- purrr::safely(jsonlite::fromJSON, otherwise = NULL, quiet = TRUE)
```

Comment:
use `safely` so that we can print out the error as a warning and still return a `tibble`. there may be better ways to do this, but this works 🤷